shared/src/vm/host_env.rs

```diff
@@ -2076,6 +2076,85 @@ where
     sentinel.set_invalid_commitment();
 }
 
+/// Verify a transaction signature
+#[allow(clippy::too_many_arguments)]
+pub fn tx_verify_tx_section_signature<MEM, DB, H, CA>(
+    env: &TxVmEnv<MEM, DB, H, CA>,
+    hash_list_ptr: u64,
+    hash_list_len: u64,
+    public_keys_map_ptr: u64,
+    public_keys_map_len: u64,
+    threshold: u8,
+    max_signatures_ptr: u64,
+    max_signatures_len: u64,
+) -> TxResult<i64>
+where
+    MEM: VmMemory,
+    DB: storage::DB + for<'iter> storage::DBIter<'iter>,
+    H: StorageHasher,
+    CA: WasmCacheAccess,
+{
+    let (hash_list, gas) = env
+        .memory
+        .read_bytes(hash_list_ptr, hash_list_len as _)
+        .map_err(|e| TxRuntimeError::MemoryError(Box::new(e)))?;
+
+    let sentinel = unsafe { env.ctx.sentinel.get() };
+    tx_charge_gas(env, gas)?;
+    let hashes = <[Hash; 1]>::try_from_slice(&hash_list)
+        .map_err(TxRuntimeError::EncodingError)?;
+
+    let (public_keys_map, gas) = env
+        .memory
+        .read_bytes(public_keys_map_ptr, public_keys_map_len as _)
+        .map_err(|e| TxRuntimeError::MemoryError(Box::new(e)))?;
+    tx_charge_gas(env, gas)?;
+    let public_keys_map =
+        namada_core::types::account::AccountPublicKeysMap::try_from_slice(
+            &public_keys_map,
+        )
+        .map_err(TxRuntimeError::EncodingError)?;
+
+    tx_charge_gas(env, gas)?;
+
+    let (max_signatures, gas) = env
+        .memory
+        .read_bytes(max_signatures_ptr, max_signatures_len as _)
+        .map_err(|e| TxRuntimeError::MemoryError(Box::new(e)))?;
+    tx_charge_gas(env, gas)?;
+    let max_signatures = Option::<u8>::try_from_slice(&max_signatures)
+        .map_err(TxRuntimeError::EncodingError)?;
+
+    let tx = unsafe { env.ctx.tx.get() };
+
+    tx_charge_gas(
+        env,
+        gas::VERIFY_TX_SIG_GAS * public_keys_map.idx_to_pk.len() as u64,
+    )?;
+    match tx.verify_signatures(
+        &hashes,
+        public_keys_map,
+        &None,
+        threshold,
+        max_signatures,
+        // This uses VpGasMeter, so we're charging the gas here manually
+        // instead
+        &mut None,
```

Comment:
Minor thing here, we can change the `verify_signatures` method of `Tx` to accept an `&mut dyn GasMetering` trait object just like we do in `fetch_or_compile` so that we can still charge gas for every single signature without the need to do it manually, if you prefer. Alternatively, since this function is only called by the `init_validator` tx, and we always require 4 valid signatures, we could also just skip charging gas in here since its cost is already included in the benchmark of the tx itself

Reply:
I've refactored it in [88578bf](https://github.com/anoma/namada/pull/2163/commits/88578bfa1b48d6587bc18c112f31611c829dea38) but without the dynamic dispatch using a lambda instead. It also moves the gas metering logic back into the relevant places in `vm/host_env.rs` instead of being done in `proto/types.rs`